Impl Add<Month> + Sub<Month> for NaiveDateTime

avantgardnerio opened the pull request:
### Thanks for contributing to chrono!

- [x] Have you added yourself and the change to the [changelog]? (Don't worry
      about adding the PR number)
- [ ] If this pull request fixes a bug, does it add a test that verifies that
      we can't reintroduce it?

[changelog]: ../CHANGELOG.md

djc commented:
This seems reasonable. I think we should probably also add `checked_add_months()` and `checked_sub_months()`, and I'd like to express the new `Add` and `Sub` implementation in terms of `NaiveDate::checked_add_months()` and `checked_sub_months()` as well, to reduce the amount of layering and make the panicking behavior more explicit.

avantgardnerio commented:
@djc I added these functions, and I think they meet the requirements listed above. I inadvertently found an edge-case that was panicking, and added a test and fixed that too (see build logs for details).

djc commented on src/naive/date.rs:
````diff
@@ -629,22 +629,28 @@ impl NaiveDate {
     /// );
     /// ```
+    ///
+    /// Returns `None` if the resulting date would be out of range.
````
I'd prefer to keep this explanation before the examples (and I think that's common style for rustdoc).

djc commented on src/naive/datetime/mod.rs:
````diff
@@ -533,6 +533,34 @@ impl NaiveDateTime {
+    /// assert_eq!(dt("2014-01-01T01:00:00").checked_add_months(m), None);
+    /// ```
+    pub fn checked_add_months(self, rhs: Months) -> Option<NaiveDateTime> {
+        self.date().checked_add_months(rhs).map(|dt| dt.and_time(self.time()))
````
I think this can be simplified to

```rust
self.date = self.date.checked_add_months(rhs)?;
```

I definitely think we should not be using the `date()` and `time()` accessors here.

djc commented on src/naive/datetime/mod.rs:
````diff
@@ -1289,6 +1345,37 @@ impl AddAssign<OldDuration> for NaiveDateTime {
+    /// assert_eq!(dt("2020-01-31T06:00:00").unwrap() + m(1), dt("2020-02-29T06:00:00").unwrap());
+    /// ```
+    fn add(self, rhs: Months) -> Self::Output {
+        let new_date = self.date() + rhs;
+        new_date.and_time(self.time())
````
```rust
Self {
    date: self.date.checked_add_months(rhs).unwrap(),
    time: self.time,
}
```

djc replied:
This doesn't seem to be resolved?

avantgardnerio replied:
Sorry, I fixed it in the other two places and forgot to fix it here. It is resolved now.

djc commented on src/naive/date.rs:
```diff
@@ -629,22 +629,28 @@ impl NaiveDate {
-        // Copy `i32::MIN` here so we don't have to do a complicated cast
-        match months.0 <= 2_147_483_648 {
+        // Copy `i32::MAX` here so we don't have to do a complicated cast
+        match months.0 <= 2_147_483_647 {
```
Ah, missed that we're only making this negative after the cast to `i32`.

djc commented on src/naive/datetime/mod.rs:
```diff
@@ -606,6 +634,34 @@ impl NaiveDateTime {
         Some(NaiveDateTime { date, time })
     }
 
+    /// Subtracts given `Months` to the current date and time.
```
Sorry, minor nit I forgot to mention: "Subtracts .. to" should use have "from" instead of "to".

esheppa commented on src/naive/datetime/mod.rs:
```diff
@@ -533,6 +533,34 @@ impl NaiveDateTime {
+    /// let dt = |s| NaiveDateTime::from_str(s).unwrap();
+    /// let ca = |dt: NaiveDateTime, m: Months| dt.checked_add_months(m).unwrap();
+    /// let m = Months::new;
+    /// assert_eq!(ca(dt("2014-01-01T01:00:00"), m(1)), dt("2014-02-01T01:00:00"));
```
I'm a little uncomfortable with this style of code (albeit similar to what is used in the private tests) being in the public docs. IMO this would look better if we just used the full functions directly, with more intermediate variables if needed for readability. 

I've set up https://github.com/avantgardnerio/chrono/pull/1 as one possibility here, but I'm not set on that option, and keen to have a discussion on the merits of each approach.

djc replied:
I agree that this makes the examples harder to understand (and somewhat unidiomatic).

avantgardnerio replied:
I think they are all fixed now. Please ping me if I missed any!

esheppa reviewed (changes requested):
Thanks @avantgardnerio - good idea to add this to `NaiveDateTime` (I'm keen to set up a PR to add it to `DateTime<Tz>` shortly after this :smile:) - I've set up a PR to your branch with a proposal for a change to the documentation, but apart from that this looks great!

avantgardnerio commented:
Thanks all! :tada:
